src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+{
+    unsigned id = static_cast<unsigned>(values.size());
+    values.push_back(hash_value{id, cur_point});
+    hash_value& val = get_value(id);
+    add_instr(id);
+    instr2_value[ins] = &val;
+    return val;
+}
+
+void horizontal_fusion_impl::process(instruction_ref ins)
+{
+    // Do not hash literals.
+    if(ins->name() == "@literal")
+        return;
+    if(instr2_hash.find(ins) != instr2_hash.end())
+    {
+        // Hash this instruction.
+        if(hash(ins) != nullptr)
+        {
+            for(auto&& output : ins->outputs())
+            {
+                instr2_hash[output] = true;
+            }
+            return;
+        }
+    }
+
+    std::unordered_map<std::string, int> op2_cnt;
+    bool hash_child = false;
+    // Do hash if at least two outputs have same operations.
+    for(auto&& output : ins->outputs())
+    {
+        const std::string& str = output->name();
+        if(op2_cnt.find(str) == op2_cnt.end())
+            op2_cnt[str] = 1;
+        else
+        {
+            op2_cnt[str] += 1;
+            hash_child = true;
+            break;
+        }
+    }
+    if(hash_child)
+    {
+        // Create a value for this instruction.
+        hash_value& value = create_value(ins);
+        value.set_root();
+        // Flag children to be hashed.
+        for(auto&& output : ins->outputs())
+        {
+            if(op2_cnt[output->name()] > 1)
+                instr2_hash[output] = true;
+        }
+    }
+}
+
+// Find the first axis that matches given dim.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, int dim)
+{
+    int ndx = 0;
+    for(auto&& size : ins->get_shape().lens())
+    {
+        if(size == dim)
+            return ndx;
+        ndx++;
+    }
+    return -1;
+}
+
+// Find axis for convolution filter's output and broadcast's input.
+int horizontal_fusion_impl::find_axis(instruction_ref ins, instruction_ref base, int base_axis)
+{
+    if(is_conv(base))
+    {
+        return get_conv_output_axis();
+    }
+    else if(ins->outputs().at(0)->name() == "broadcast")
+    {
+        if(!match_dim(base, ins->outputs().at(0), -1))
+            MIGRAPHX_THROW("Unmatched output");
+        int dim = base->get_shape().lens().at(base_axis);
+        return find_axis(ins, dim);
+    }
+    else
+        return base_axis;
+}
+
+// Check whether ins1 and ins2 match in all dimensions excluding axis.
+bool horizontal_fusion_impl::match_dim(instruction_ref ins1, instruction_ref ins2, int axis)
+{
+    auto lens1 = ins1->get_shape().lens();
+    auto lens2 = ins2->get_shape().lens();
+    if(lens1.size() != lens2.size())
+        return false;
+    int ndx = 0;
+    for(auto&& size : lens1)
+    {
+        if((size != lens2.at(ndx)) && (ndx != axis))
+            return false;
+        ndx++;
+    }
+    return true;
+}
+
+bool horizontal_fusion_impl::compare_inputs(std::vector<instruction_ref>& input1,
+                                            std::vector<instruction_ref>& input2,
```

Comment:
These parameters can be passed as const ref.

Reply:
fixed

Reply:
fixed.